File: backend/app/schemas/credential.py
"""Pydantic schemas for the credential API."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator

from app.schemas.markers import check_reserved_marker

# ---- Catalog ---------------------------------------------------------------


class CredentialFieldSchema(BaseModel):
    name: str
    display_name: str
    kind: str
    default: Any = None
    required: bool = False
    description: str | None = None
    options: list[dict[str, Any]] = Field(default_factory=list)
    placeholder: str | None = None
    type_options: dict[str, Any] = Field(default_factory=dict)
    display_options: dict[str, Any] = Field(default_factory=dict)


class CredentialDefinitionSchema(BaseModel):
    key: str
    display_name: str
    icon_id: str | None = None
    documentation_url: str | None = None
    category: str = "general"
    extends: list[str] = Field(default_factory=list)
    properties: list[CredentialFieldSchema] = Field(default_factory=list)
    has_test: bool = False
    has_oauth: bool = False


# ---- CRUD ------------------------------------------------------------------


class CredentialCreate(BaseModel):
    definition_key: str
    name: str
    data: dict[str, Any]
    is_shared: bool = False

    @field_validator("name")
    @classmethod
    def _validate_name(cls, value: str) -> str:
        check_reserved_marker(value, "name")
        return value

    def normalized_name(self) -> str:
        return self.name


class CredentialUpdate(BaseModel):
    name: str | None = None
    data: dict[str, Any] | None = None
    is_shared: bool | None = None
    status: Literal["active", "disabled", "expired"] | None = None

    @field_validator("name")
    @classmethod
    def _validate_name(cls, value: str | None) -> str | None:
        return check_reserved_marker(value, "name")


class CredentialResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    definition_key: str
    name: str
    field_keys: list[str]
    is_shared: bool
    status: str
    key_id: str
    last_used_at: datetime | None = None
    last_tested_at: datetime | None = None
    last_test_result: dict[str, Any] | None = None
    created_at: datetime
    updated_at: datetime


# ---- Test ------------------------------------------------------------------


class CredentialTestResponse(BaseModel):
    success: bool
    http_status: int | None = None
    message: str
    details: dict[str, Any] = Field(default_factory=dict)


class PreviewTestRequest(BaseModel):
    definition_key: str
    data: dict[str, Any]


# ---- Audit log -------------------------------------------------------------


class CredentialAuditLogResponse(BaseModel):
    id: uuid.UUID
    credential_id: uuid.UUID
    actor_user_id: uuid.UUID | None = None
    action: str
    source: str
    ip: str | None = None
    user_agent: str | None = None
    error: str | None = None
    log_metadata: dict[str, Any] | None = None
    created_at: datetime


# ---- OAuth2 helpers --------------------------------------------------------


class OAuth2AuthStartResponse(BaseModel):
    authorization_url: str
    state: str

File: frontend/src/lib/types/credential.ts
// Credential domain types — mirrors backend `app/schemas/credential.py`.
// Field, definition, instance, audit log, OAuth2 helpers.

export type FieldKind =
  | 'string'
  | 'password'
  | 'number'
  | 'select'
  | 'multiline'
  | 'json'
  | 'oauth_button'
  | 'toggle'
  | 'collection'

export interface FieldOption {
  name?: string
  value: string | number | boolean
  description?: string
}

export interface FieldTypeOptions {
  password?: boolean
  multiline?: boolean
  rows?: number
  expirable?: boolean
  min?: number
  max?: number
  step?: number
  regex?: string
  min_length?: number
  max_length?: number
}

export interface FieldDisplayOptions {
  // condition: parent field name -> array of accepted values
  show?: Record<string, Array<string | number | boolean>>
  hide?: Record<string, Array<string | number | boolean>>
}

export interface FieldDef {
  name: string
  display_name: string
  kind: FieldKind
  default?: unknown
  required?: boolean
  description?: string | null
  options?: FieldOption[]
  placeholder?: string | null
  type_options?: FieldTypeOptions
  display_options?: FieldDisplayOptions
  // Tool parameters only: filled in by the agent at call time, hidden in the
  // tool-creation form. Backend exposes the field on the LLM args schema.
  runtime_only?: boolean
}

export interface CredentialDefinition {
  key: string
  display_name: string
  icon_id?: string | null
  documentation_url?: string | null
  category: string
  extends: string[]
  properties: FieldDef[]
  has_test: boolean
  has_oauth: boolean
}

export type CredentialStatus = 'active' | 'auth_needed' | 'expired' | 'disabled' | 'unknown'

export interface Credential {
  id: string
  user_id: string
  definition_key: string
  name: string
  field_keys: string[]
  is_shared: boolean
  // True for operator-managed system credentials (Fix Agent / builder /
  // image generation). User-facing pickers MUST hide these.
  is_system?: boolean
  status: CredentialStatus | string
  key_id: string
  last_used_at: string | null
  last_tested_at: string | null
  last_test_result: Record<string, unknown> | null
  created_at: string
  updated_at: string
}

export interface CredentialCreateRequest {
  definition_key: string
  name: string
  data: Record<string, unknown>
  is_shared?: boolean
}

export interface CredentialUpdateRequest {
  name?: string
  data?: Record<string, unknown>
  is_shared?: boolean
  status?: 'active' | 'disabled' | 'expired'
}

export interface CredentialTestResult {
  success: boolean
  http_status?: number | null
  message: string
  details?: Record<string, unknown>
}

export interface CredentialAuditLog {
  id: string
  credential_id: string
  actor_user_id: string | null
  action: string
  source: string
  ip: string | null
  user_agent: string | null
  error: string | null
  log_metadata: Record<string, unknown> | null
  created_at: string
}

export interface OAuth2AuthStartResponse {
  authorization_url: string
  state: string
}

export interface PreviewTestRequest {
  definition_key: string
  data: Record<string, unknown>
}
